testutil/capturelog.go

```diff
@@ -4,79 +4,129 @@ import (
 // CaptureLogger defines a logging structure for plugins.
 type CaptureLogger struct {
-	Name   string // Name is the plugin name, will be printed in the `[]`.
-	errors []string
+	T        *testing.T
+	Name     string // Name is the plugin name, will be printed in the `[]`.
+	messages []logMessage
 	sync.Mutex
 }
 
-// Errorf logs an error message, patterned after log.Printf.
-func (l *CaptureLogger) Errorf(format string, args ...interface{}) {
-	s := fmt.Sprintf("E! ["+l.Name+"] "+format, args...)
+func (l *CaptureLogger) msgString(msg logMessage) string {
+	return fmt.Sprintf("%c! [%s] %s", msg.level, l.Name, msg.message)
+}
+func (l *CaptureLogger) logMsg(msg logMessage) {
 	l.Lock()
-	l.errors = append(l.errors, s)
+	l.messages = append(l.messages, msg)
+	l.T.Log(l.msgString(msg))
 	l.Unlock()
-	log.Print(s)
+}
+
+func (l *CaptureLogger) logf(level byte, format string, args ...any) {
+	l.logMsg(logMessage{level, fmt.Sprintf(format, args...)})
+}
+
+func (l *CaptureLogger) loga(level byte, args ...any) {
+	l.logMsg(logMessage{level, fmt.Sprint(args...)})
+}
+
+// Errorf logs an error message, patterned after log.Printf.
+func (l *CaptureLogger) Errorf(format string, args ...interface{}) {
+	l.logf(logLevelError, format, args...)
 }
 
 // Error logs an error message, patterned after log.Print.
 func (l *CaptureLogger) Error(args ...interface{}) {
-	s := fmt.Sprint(append([]interface{}{"E! [" + l.Name + "] "}, args...)...)
-	l.Lock()
-	l.errors = append(l.errors, s)
-	l.Unlock()
-	log.Print(s)
+	l.loga(logLevelError, args...)
 }
 
 // Debugf logs a debug message, patterned after log.Printf.
 func (l *CaptureLogger) Debugf(format string, args ...interface{}) {
-	log.Printf("D! ["+l.Name+"] "+format, args...)
+	l.logf(logLevelDebug, format, args...)
 }
 
 // Debug logs a debug message, patterned after log.Print.
 func (l *CaptureLogger) Debug(args ...interface{}) {
-	log.Print(append([]interface{}{"D! [" + l.Name + "] "}, args...)...)
+	l.loga(logLevelDebug, args...)
 }
 
 // Warnf logs a warning message, patterned after log.Printf.
 func (l *CaptureLogger) Warnf(format string, args ...interface{}) {
 	log.Printf("W! ["+l.Name+"] "+format, args...)
+	l.logf(logLevelWarn, format, args...)
 }
 
 // Warn logs a warning message, patterned after log.Print.
 func (l *CaptureLogger) Warn(args ...interface{}) {
-	log.Print(append([]interface{}{"W! [" + l.Name + "] "}, args...)...)
+	l.loga(logLevelWarn, args...)
 }
 
 // Infof logs an information message, patterned after log.Printf.
 func (l *CaptureLogger) Infof(format string, args ...interface{}) {
-	log.Printf("I! ["+l.Name+"] "+format, args...)
+	l.logf(logLevelInfo, format, args...)
 }
 
 // Info logs an information message, patterned after log.Print.
 func (l *CaptureLogger) Info(args ...interface{}) {
-	log.Print(append([]interface{}{"I! [" + l.Name + "] "}, args...)...)
+	l.loga(logLevelInfo, args...)
 }
 
-func (l *CaptureLogger) Errors() []string {
+func (l *CaptureLogger) Messages() []string {
```

Comment:
TBH now that we have the level separated in a struct we should really return `[]LogMessage` as this allows for easier filtering in case the test is only interested in e.g. the warn messages.